Preliminary MSIL-driven Roslyn diagnostics analyzer support (for symbol actions only).

michaelcfanning opened the pull request:
@nguerrera @BillyONeal @lgolding @tikiwanMS @bobschumaker @davidknise

msftclas commented:
Hi __@michaelcfanning__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Michael Fanning (TWC)). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

nguerrera commented on src/BinSkim.Driver/RoslynAnalyzer/RoslynSymbolVisitor.cs:
```diff
@@ -0,0 +1,47 @@
+        {
+            foreach (var member in symbol.GetMembers()) { Visit(member); }
+
+            base.VisitNamespace(symbol);
```
Nit: inconsistent spacing between foreach and base call vs. the others.

michaelcfanning replied:
fixed

michaelcfanning commented on src/BinSkim.Driver/AnalyzeCommand.cs:
```diff
@@ -261,58 +270,114 @@ private HashSet<IBinarySkimmer> CreateSkimmers(IMessageLogger<BinaryAnalyzerCont
+        {
+            ILDiagnosticsAnalyzer roslynAnalyzer = new ILDiagnosticsAnalyzer();
+
+            foreach (string analyzerFilePath in roslynAnalyzerFilePaths)
```
Note that we reload all analyzers for each analysis target, which entails reflecting over the analyzer set again and again. This could be made more efficient.

nguerrera replied:
Since we use a fresh ILDiagnosticsAnalyzer and only call Analyze once, my earlier comment about bleeding/duplicated registrations is mostly academic.

Maybe just comment that we don't currently support more than one Load/Analyze pair in ILDiagnosticsAnalyzer, but better to either multiple analyzes per instance correctly or change the ILDiagnosticsAnalyzer interface to prevent it.

michaelcfanning commented on src/BinSkim.Driver/AnalyzeCommand.cs:
```diff
@@ -261,58 +270,114 @@ private HashSet<IBinarySkimmer> CreateSkimmers(IMessageLogger<BinaryAnalyzerCont
+                );
             }
+
+            roslynAnalyzer.Analyze(assemblyFilePath, diagnostic =>
```
Note that this code is not tested for exceptional cases. I will file an issue on this (and on the previous issue related to reloading analyzers over and over)

michaelcfanning commented on src/BinSkim.Driver/AnalyzeCommand.cs:
```diff
@@ -261,58 +272,122 @@ private HashSet<IBinarySkimmer> CreateSkimmers(IMessageLogger<BinaryAnalyzerCont
+        {
+            ILDiagnosticsAnalyzer roslynAnalyzer = new ILDiagnosticsAnalyzer();
+
+            if (_globalRoslynAnalysisContext == null)
```
ok, now we retain a global analysis context object that is initialized once only. we push this onto new ILDiagnosticsAnalyzer instances rather than reloading from analyzer paths. Did I get all this right? I will defer the end compilation action to a future change: because this callback can report diagnostics, I will need to plumb through a mechanism to log through typical channels.

michaelcfanning commented:
@nguerrera  added compilation end callbacks on both global analysis context and compilation start context. updated brain-dead unit tests to analyze multiple files and to fire messages for relevant compilation end callbacks (to verify we observe the expected set). cleaned up the ILDiagnosticsAnalyzer use pattern. we now have two Create factory methods that provide for a single point of initialization. You can load individual analyzers but only against a context object (rather than an ILDiagnosticsAnalyzer instance).

michaelcfanning commented:
also, going ahead with the merge on this (but I hope you will take a good look at it) as I'd like to move on to some work updating the output format (which is likely to leak into many places).

nguerrera commented on src/BinSkim.Driver/RoslynAnalyzer/ILDiagnosticsAnalyzer.cs:
I would pass these down to `Analyze(ISymbol, ...)` and avoid the field that we have to null out. It will get captured with the right lifetime in the `symbol => Analyze(...)` closure.

nguerrera commented on src/BinSkim.Driver/RoslynAnalyzer/RoslynCompilationStartAnalysisContext.cs:
```diff
@@ -0,0 +1,57 @@
+    /// </summary>
+    internal sealed class RoslynCompilationStartAnalysisContext : CompilationStartAnalysisContext
+    {
+        public ActionMap<SymbolAnalysisContext, SymbolKind> SymbolActions { get; set; }
```
Why is this publicly settable?

nguerrera commented on src/BinSkim.Driver/RoslynAnalyzer/RoslynCompilationStartAnalysisContext.cs:
```diff
@@ -0,0 +1,57 @@
+    {
+        public ActionMap<SymbolAnalysisContext, SymbolKind> SymbolActions { get; set; }
+
+        public Action<CompilationAnalysisContext> CompilationEndActions { get; set; }
```
Ditto. Why public setter? I would expect these to be ReadOnly with RegisterXxx the only way to add new actions.

nguerrera commented:
LGTM with a few minor comments on changes since the last iteration I reviewed.
